File: options_lab/ic.py
"""Panel information coefficient, with its null band and its controls attached.

An IC reported alone is not evidence. Two of the most widely published option
"signals" fail exactly here: PCR's raw IC of -0.33 at 60 minutes collapses to
-0.086 +/- 0.057 once the index level is controlled, and max pain's +0.58 is
entirely a monotone transform of minus-the-index - a control containing no
option data at all ("distance to the day's open") beats it, +0.644 vs +0.584.

So `ICRow.format()` raises unless controls were supplied. The omission is the
failure mode, so the type refuses to make it.

Significance is day-clustered. Every observation inside a session shares one
underlying path; the iid interval is reported beside the clustered one so the
gap is visible rather than assumed away.
"""
from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np
import pandas as pd

NULL_DRAWS = 20          # a band well inside the decision width; 200 buys a decimal
DECILES = 10


class UncontrolledIC(RuntimeError):
    """A raw IC has no reportable meaning without its partial beside it."""


def _spearman(a: pd.Series, b: pd.Series) -> float:
    ok = a.notna() & b.notna()
    a, b = a[ok], b[ok]
    if len(a) < 3 or a.nunique() < 2 or b.nunique() < 2:
        return float("nan")
    return float(a.rank().corr(b.rank()))


# Below this share of the original variance, a residual is numerical dust and
# any correlation computed on it is an artefact of the solver, not information.
DEGENERATE_VAR_RATIO = 1e-12


def _residualise(y: pd.Series, controls: pd.DataFrame) -> pd.Series:
    """Least-squares residual of y on controls plus an intercept.

    Returns all-NaN when the controls explain y essentially completely. Two such
    residual vectors are proportional floating-point dust, and their rank
    correlation comes out near +/-1 - a spurious partial IC precisely where the
    honest answer is "the controls already explain this feature".
    """
    X = np.column_stack([np.ones(len(y)), controls.to_numpy(dtype="float64")])
    yv = y.to_numpy(dtype="float64")
    beta, *_ = np.linalg.lstsq(X, yv, rcond=None)
    resid = yv - X @ beta

    var_y = float(np.var(yv))
    if var_y > 0 and float(np.var(resid)) / var_y < DEGENERATE_VAR_RATIO:
        return pd.Series(np.nan, index=y.index)
    return pd.Series(resid, index=y.index)


@dataclass(frozen=True)
class ICRow:
    feature: str
    horizon: int
    n_obs: int
    n_sessions: int
    ic: float
    daily_ic_mean: float
    daily_ic_lo: float
    daily_ic_hi: float
    ic_iid_lo: float
    ic_iid_hi: float
    null_ic_hi: float
    decile_spread_pts: float
    breakeven_pts: float
    edge_over_cost: float
    partial_ic: float = float("nan")
    controlled: bool = field(default=False)
    # A directional trade earns its own decile's mean, not D10 minus D1.
    # Dividing the full long-short spread by one round trip overstates a
    # single-leg strategy by ~2x.
    single_leg_edge_pts: float = float("nan")
    null_is_daily: bool = True
    # Sessions that actually yielded an IC. A session with no rank variance
    # is dropped, so daily_ic_mean and its interval rest on FEWER sessions
    # than n_sessions. Reporting only the total overstates the evidence.
    n_sessions_measured: int = 0

    def format(self) -> str:
        if not self.controlled:
            raise UncontrolledIC(
                f"{self.feature}: refusing to report a raw IC of {self.ic:.4f} with "
                "no partial beside it. Supply controls - at minimum the index's "
                "contemporaneous return and the lagged return. PCR and max pain "
                "both look real until this is done."
            )
        return (
            f"{self.feature} h={self.horizon} n={self.n_obs} "
            f"sessions={self.n_sessions_measured}/{self.n_sessions} "
            f"ic={self.ic:+.4f} "
            f"partial_ic={self.partial_ic:+.4f} "
            f"daily[{self.daily_ic_lo:+.4f},{self.daily_ic_hi:+.4f}] "
            f"iid[{self.ic_iid_lo:+.4f},{self.ic_iid_hi:+.4f}] "
            f"null_hi={self.null_ic_hi:.4f} "
            f"spread={self.decile_spread_pts:+.2f}pts "
            f"leg={self.single_leg_edge_pts:+.2f}pts "
            f"breakeven={self.breakeven_pts:.2f}pts "
            f"edge_over_cost={self.edge_over_cost:.2f}x"
            f"{'' if self.monotone else ' [NON-MONOTONE: tails disagree with IC]'}"
        )

    @property
    def monotone(self) -> bool:
        """Do the tails agree with the rank correlation?

        The decile spread and the IC are different measurements, and they can
        disagree - a non-monotone relationship where the extremes behave
        opposite to the overall ranking. When they disagree, `edge_over_cost`
        is a magnitude with no coherent direction to trade, so it must not be
        read as edge.
        """
        if np.isnan(self.ic) or np.isnan(self.decile_spread_pts):
            return False
        return np.sign(self.ic) == np.sign(self.decile_spread_pts)

    @property
    def clears_cost(self) -> bool:
        return self.edge_over_cost >= 2.0 and self.monotone

    @property
    def beats_null(self) -> bool:
        """Compared against the DAILY-mean null, matching daily_ic_lo/hi.

        These were previously mismatched: the null band was built from pooled
        Spearmans while the interval was built from per-session ones, so the
        two scored different statistics and were nonetheless compared.
        """
        return abs(self.daily_ic_mean) > self.null_ic_hi


def panel_ic(
    *,
    feature: pd.Series,
    forward: pd.Series,
    sessions: pd.Series,
    controls: pd.DataFrame | None = None,
    name: str = "feature",
    horizon: int = 1,
    breakeven_pts: float = 1.0,
    null_draws: int = NULL_DRAWS,
    seed: int = 0,
) -> ICRow:
    """IC of `feature` against `forward`, clustered by `sessions`."""
    n = len(feature)
    if not (len(forward) == len(sessions) == n):
        raise ValueError(
            f"length mismatch: feature={n} forward={len(forward)} "
            f"sessions={len(sessions)} - align before measuring, never truncate"
        )

    feature = feature.reset_index(drop=True)
    forward = forward.reset_index(drop=True)
    sessions = sessions.reset_index(drop=True)

    overall = _spearman(feature, forward)

    # Day-clustered: one IC per session, then a mean and a percentile interval
    # over sessions. This is the interval that respects the shared path.
    per_day = pd.Series(
        {s: _spearman(feature[m], forward[m])
         for s, m in feature.groupby(sessions).groups.items()
         for m in [sessions == s]}
    ).dropna()
    daily_mean = float(per_day.mean()) if len(per_day) else float("nan")
    if len(per_day) >= 2:
        se = float(per_day.std(ddof=1) / np.sqrt(len(per_day)))
        daily_lo, daily_hi = daily_mean - 1.96 * se, daily_mean + 1.96 * se
    else:
        daily_lo = daily_hi = float("nan")

    # The iid interval, reported only so the gap to the clustered one is visible.
    se_iid = 1.0 / np.sqrt(max(n - 3, 1))
    ic_iid_lo, ic_iid_hi = overall - 1.96 * se_iid, overall + 1.96 * se_iid

    # Null: permute the feature's SIGN within each session. Magnitude
    # distribution and time-of-day profile survive; only direction is destroyed.
    rng = np.random.default_rng(seed)
    draws = []
    for _ in range(null_draws):
        flipped = feature.copy()
        for _s, idx in feature.groupby(sessions).groups.items():
            signs = rng.choice([-1.0, 1.0], size=len(idx))
            flipped.loc[idx] = feature.loc[idx].to_numpy() * signs
        per_day_null = [_spearman(flipped[m], forward[m])
                        for m in (sessions == s for s in sessions.unique())]
        draws.append(abs(np.nanmean(per_day_null)))
    null_hi = float(np.nanpercentile(draws, 95)) if draws else float("nan")

    # Decile spread, in the same units as `forward`.
    try:
        buckets = pd.qcut(feature.rank(method="first"), DECILES, labels=False)
        by = forward.groupby(buckets).mean()
        spread = float(by.iloc[-1] - by.iloc[0])
    except (ValueError, IndexError):
        spread = float("nan")

    partial = float("nan")
    if controls is not None and not controls.empty:
        c = controls.reset_index(drop=True)
        partial = _spearman(_residualise(feature, c), _residualise(forward, c))

    single_leg = spread / 2.0 if not np.isnan(spread) else float("nan")

    return ICRow(
        feature=name, horizon=horizon, n_obs=n, n_sessions=int(sessions.nunique()),
        n_sessions_measured=int(len(per_day)),
        ic=overall, daily_ic_mean=daily_mean, daily_ic_lo=daily_lo,
        daily_ic_hi=daily_hi, ic_iid_lo=ic_iid_lo, ic_iid_hi=ic_iid_hi,
        null_ic_hi=null_hi, decile_spread_pts=spread, breakeven_pts=breakeven_pts,
        edge_over_cost=(abs(single_leg) / breakeven_pts
                        if breakeven_pts else float("nan")),
        partial_ic=partial, controlled=controls is not None and not controls.empty,
        single_leg_edge_pts=single_leg,
    )

File: android/engine/src/main/kotlin/com/optionslab/engine/Ic.kt
package com.optionslab.engine

import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Panel information coefficient, with its null band and its controls attached.
 * Port of `ic.py` and the `ic` command.
 *
 * An IC reported alone is not evidence: [IcRow.format] refuses to print a raw
 * IC without its partial beside it, because that omission is exactly how PCR
 * and max pain look real.
 */
object Ic {
    const val NULL_DRAWS = 20
    const val DECILES = 10
    const val DEGENERATE_VAR_RATIO = 1e-12
    val HORIZONS = listOf(1, 5, 15, 30)
    val LOTS = mapOf("NIFTY" to 65, "BANKNIFTY" to 30)

    class UncontrolledIC(msg: String) : IllegalStateException(msg)

    data class IcRow(
        val feature: String, val horizon: Int, val nObs: Int, val nSessions: Int,
        val ic: Double, val dailyIcMean: Double, val dailyIcLo: Double, val dailyIcHi: Double,
        val icIidLo: Double, val icIidHi: Double, val nullIcHi: Double,
        val decileSpreadPts: Double, val breakevenPts: Double, val edgeOverCost: Double,
        val partialIc: Double, val controlled: Boolean, val singleLegEdgePts: Double,
        val nSessionsMeasured: Int,
    ) {
        val monotone: Boolean get() =
            !ic.isNaN() && !decileSpreadPts.isNaN() && Math.signum(ic) == Math.signum(decileSpreadPts)
        val clearsCost: Boolean get() = edgeOverCost >= 2.0 && monotone
        val beatsNull: Boolean get() = abs(dailyIcMean) > nullIcHi

        fun format(): String {
            if (!controlled) throw UncontrolledIC(
                "$feature: refusing to report a raw IC of %.4f with no partial beside it.".format(ic))
            return ("$feature h=$horizon n=$nObs sessions=$nSessionsMeasured/$nSessions ic=%+.4f partial_ic=%+.4f " +
                "daily[%+.4f,%+.4f] iid[%+.4f,%+.4f] null_hi=%.4f spread=%+.2fpts leg=%+.2fpts breakeven=%.2fpts " +
                "edge_over_cost=%.2fx").format(ic, partialIc, dailyIcLo, dailyIcHi, icIidLo, icIidHi, nullIcHi,
                decileSpreadPts, singleLegEdgePts, breakevenPts, edgeOverCost) +
                if (monotone) "" else " [NON-MONOTONE: tails disagree with IC]"
        }
    }

    /** Least-squares residual of y on controls plus an intercept (all-NaN if degenerate). */
    fun residualise(y: DoubleArray, controls: List<DoubleArray>): DoubleArray {
        val n = y.size
        val k = controls.size + 1
        val x = Array(n) { i -> DoubleArray(k) { j -> if (j == 0) 1.0 else controls[j - 1][i] } }
        val xtx = Array(k) { DoubleArray(k) }
        val xty = DoubleArray(k)
        for (i in 0 until n) for (a in 0 until k) {
            xty[a] += x[i][a] * y[i]
            for (b in 0 until k) xtx[a][b] += x[i][a] * x[i][b]
        }
        val beta = solve(xtx, xty)
        val resid = DoubleArray(n) { i -> y[i] - (0 until k).sumOf { x[i][it] * beta[it] } }
        val varY = variance(y)
        if (varY > 0 && variance(resid) / varY < DEGENERATE_VAR_RATIO) return DoubleArray(n) { Double.NaN }
        return resid
    }

    private fun variance(v: DoubleArray): Double {
        if (v.isEmpty()) return 0.0
        val m = v.average()
        return v.sumOf { (it - m) * (it - m) } / v.size
    }

    /** Gaussian elimination with partial pivoting; singular columns get a zero coefficient (min-norm-like). */
    private fun solve(a0: Array<DoubleArray>, b0: DoubleArray): DoubleArray {
        val n = b0.size
        val a = Array(n) { a0[it].copyOf() }
        val b = b0.copyOf()
        val piv = IntArray(n) { -1 }
        var row = 0
        for (col in 0 until n) {
            var best = row
            for (r in row until n) if (abs(a[r][col]) > abs(a[best][col])) best = r
            if (row >= n || abs(a[best][col]) < 1e-12 * (1 + abs(a[best][col]))) continue
            a[row] = a[best].also { a[best] = a[row] }
            b[row] = b[best].also { b[best] = b[row] }
            for (r in 0 until n) if (r != row) {
                val f = a[r][col] / a[row][col]
                if (f != 0.0) { for (c in col until n) a[r][c] -= f * a[row][c]; b[r] -= f * b[row] }
            }
            piv[col] = row
            row++
        }
        return DoubleArray(n) { c -> if (piv[c] >= 0) b[piv[c]] / a[piv[c]][c] else 0.0 }
    }

    fun panelIc(
        feature: DoubleArray, forward: DoubleArray, sessions: IntArray, controls: List<DoubleArray>?,
        name: String, horizon: Int, breakevenPts: Double, nullDraws: Int = NULL_DRAWS, seed: Long = 0,
    ): IcRow {
        val n = feature.size
        require(forward.size == n && sessions.size == n) {
            "length mismatch: feature=$n forward=${forward.size} sessions=${sessions.size} - align before measuring, never truncate"
        }
        val overall = Stats.spearman(feature, forward)
        val groups = sessions.indices.groupBy { sessions[it] }
        fun sub(a: DoubleArray, idx: List<Int>) = DoubleArray(idx.size) { a[idx[it]] }
        val perDay = groups.values.map { idx -> Stats.spearman(sub(feature, idx), sub(forward, idx)) }.filter { !it.isNaN() }
        val dailyMean = if (perDay.isNotEmpty()) perDay.average() else Double.NaN
        var lo = Double.NaN
        var hi = Double.NaN
        if (perDay.size >= 2) {
            val se = Stats.std(perDay) / sqrt(perDay.size.toDouble())
            lo = dailyMean - 1.96 * se; hi = dailyMean + 1.96 * se
        }
        val seIid = 1.0 / sqrt(maxOf(n - 3, 1).toDouble())

        // Null: permute the feature's SIGN within each session.
        val rng = java.util.Random(seed)
        val draws = ArrayList<Double>()
        repeat(nullDraws) {
            val flipped = DoubleArray(n) { feature[it] * (if (rng.nextBoolean()) 1.0 else -1.0) }
            val nulls = groups.values.map { idx -> Stats.spearman(sub(flipped, idx), sub(forward, idx)) }.filter { !it.isNaN() }
            if (nulls.isNotEmpty()) draws += abs(nulls.average())
        }
        val nullHi = if (draws.isNotEmpty()) Stats.percentile(draws, 95.0) else Double.NaN

        val spread = decileSpread(feature, forward)
        var partial = Double.NaN
        val controlled = controls != null && controls.isNotEmpty()
        if (controlled) partial = Stats.spearman(residualise(feature, controls!!), residualise(forward, controls))
        val single = if (spread.isNaN()) Double.NaN else spread / 2.0
        return IcRow(
            feature = name, horizon = horizon, nObs = n, nSessions = groups.size,
            ic = overall, dailyIcMean = dailyMean, dailyIcLo = lo, dailyIcHi = hi,
            icIidLo = overall - 1.96 * seIid, icIidHi = overall + 1.96 * seIid, nullIcHi = nullHi,
            decileSpreadPts = spread, breakevenPts = breakevenPts,
            edgeOverCost = if (breakevenPts != 0.0) abs(single) / breakevenPts else Double.NaN,
            partialIc = partial, controlled = controlled, singleLegEdgePts = single,
            nSessionsMeasured = perDay.size,
        )
    }

    /** pd.qcut(feature.rank(method="first"), 10): mean forward of the top decile minus the bottom. */
    fun decileSpread(feature: DoubleArray, forward: DoubleArray): Double {
        val n = feature.size
        if (n < DECILES) return Double.NaN
        val order = feature.indices.sortedWith(compareBy({ feature[it] }, { it }))
        val rank = DoubleArray(n)
        order.forEachIndexed { r, i -> rank[i] = r + 1.0 }
        val sortedRanks = (1..n).map { it.toDouble() }
        val edges = (0..DECILES).map { Stats.percentile(sortedRanks, it * 100.0 / DECILES) }
        val sums = DoubleArray(DECILES)
        val counts = IntArray(DECILES)
        for (i in 0 until n) {
            var b = 0
            while (b < DECILES - 1 && rank[i] > edges[b + 1]) b++
            if (!forward[i].isNaN()) { sums[b] += forward[i]; counts[b]++ }
        }
        if (counts[0] == 0 || counts[DECILES - 1] == 0) return Double.NaN
        return sums[DECILES - 1] / counts[DECILES - 1] - sums[0] / counts[0]
    }

    /** One session's (flow, flow_norm, forward, index_ret, lag_ret) aligned on the index grid. */
    class Panel(val flow: DoubleArray, val flowNorm: DoubleArray, val forward: DoubleArray, val indexRet: DoubleArray, val lagRet: DoubleArray) {
        val size get() = flow.size
    }

    fun sessionPanel(day: Session, horizon: Int): Panel? {
        val index = day.index ?: return null
        val chain = day.options
        if (index.size < horizon + 5 || chain.isEmpty()) return null
        val raw = Flow.signedFlow(chain)
        val norm = Flow.signedFlow(chain, normalise = true)
        val c = index.close
        val n = c.size
        val f = ArrayList<Double>(); val fn = ArrayList<Double>(); val fw = ArrayList<Double>()
        val ir = ArrayList<Double>(); val lr = ArrayList<Double>()
        for (i in 0 until n) {
            if (i + horizon >= n || i - horizon < 0 || i < 1) continue
            val m = index.minutes[i]
            f += raw[m] ?: 0.0; fn += norm[m] ?: 0.0
            fw += c[i + horizon] - c[i]
            ir += c[i] - c[i - horizon]
            lr += c[i] - c[i - 1]
        }
        if (f.size <= horizon) return null
        return Panel(f.toDoubleArray(), fn.toDoubleArray(), fw.toDoubleArray(), ir.toDoubleArray(), lr.toDoubleArray())
    }

    /** Median traded premium of the 20 most active contracts - the cost denominator. */
    fun atmPremium(day: Session): Double? {
        val chain = day.options.filter { it.volume != null }
        val active = chain.map { s -> s to (0 until s.size).filter { s.volume!![it] > 0 } }.filter { it.second.isNotEmpty() }
        if (active.isEmpty()) return null
        val busiest = active.sortedByDescending { (s, idx) -> idx.sumOf { s.volume!![it] } }.take(20)
        return Stats.median(busiest.flatMap { (s, idx) -> idx.map { s.close[it] } })
    }

    data class IcResult(
        val underlying: String, val sessions: List<java.time.LocalDate>, val medianPremium: Double,
        val lot: Int, val regime: String, val breakeven: Double, val rows: List<IcRow>,
    ) {
        val cleared: List<IcRow> get() = rows.filter { it.clearsCost && it.beatsNull }
    }

    /**
     * The `ic` command. Days are consumed as a sequence and never held
     * together: each is reduced to its small per-horizon panels as it streams.
     */
    fun measure(underlying: String, days: Sequence<Session>, regime: String, progress: (Int) -> Unit = {}): IcResult {
        val lot = LOTS.getValue(underlying)
        val prem = ArrayList<Double>()
        val seen = ArrayList<java.time.LocalDate>()
        val panels = HORIZONS.associateWith { ArrayList<Pair<Int, Panel>>() }
        days.forEachIndexed { i, d ->
            seen += d.day
            atmPremium(d)?.let { prem += it }
            for (h in HORIZONS) sessionPanel(d, h)?.let { panels.getValue(h) += i to it }
            progress(i + 1)
        }
        require(seen.isNotEmpty()) { "no sessions to measure. Run the harvester first." }
        require(prem.isNotEmpty()) { "no traded option premiums found; cannot express cost in index points." }
        val median = Stats.median(prem)
        val breakeven = Costs.breakevenIndexPoints(median, lot, 1, regime, 0.5)
        val rows = ArrayList<IcRow>()
        for (h in HORIZONS) {
            val ps = panels.getValue(h)
            if (ps.isEmpty()) continue
            val sess = ps.flatMap { (i, p) -> List(p.size) { i } }.toIntArray()
            fun cat(sel: (Panel) -> DoubleArray) = ps.flatMap { sel(it.second).asList() }.toDoubleArray()
            val fwd = cat { it.forward }
            val controls = listOf(cat { it.indexRet }, cat { it.lagRet })
            rows += panelIc(cat { it.flow }, fwd, sess, controls, "flow", h, breakeven)
            rows += panelIc(cat { it.flowNorm }, fwd, sess, controls, "flow_norm", h, breakeven)
        }
        return IcResult(underlying, seen, median, lot, regime, breakeven, rows)
    }
}
